fix payload validation when creating projects from custom grids [MARXAN-1289]

hotzevzl opened the pull request:
https://vizzuality.atlassian.net/browse/MARXAN-1289

hotzevzl commented:
@aagm the `irregular` planning unit grid shape is not really used anywhere in the code anymore; "irregular" is _mentioned_ in a unit test, but then we use `PlanningUnitGridShape.fromShapefile` in the code.

db-wise, we have this:
```
-- geodb
marxan-geo-api=# \dT+ shape_type
                                               List of data types
 Schema |    Name    | Internal name | Size |    Elements    |     Owner      | Access privileges | Description
--------+------------+---------------+------+----------------+----------------+-------------------+-------------
 public | shape_type | shape_type    | 4    | square        +| marxan-geo-api |                   |
        |            |               |      | hexagon       +|                |                   |
        |            |               |      | irregular     +|                |                   |
        |            |               |      | from_shapefile |                |                   |
(1 row)

marxan-geo-api=# \d+ planning_units_geom
                                                                   Partitioned table "public.planning_units_geom"
    Column     |       Type       | Collation | Nullable |                              Default                               | Storage  | Compression | Stats target | Description
---------------+------------------+-----------+----------+--------------------------------------------------------------------+----------+-------------+--------------+-------------
 id            | uuid             |           | not null | gen_random_uuid()                                                  | plain    |             |              |
 the_geom      | geometry         |           | not null |                                                                    | main     |             |              |
 type          | shape_type       |           | not null |                                                                    | plain    |             |              |
 size          | integer          |           |          |                                                                    | plain    |             |              |
 area          | double precision |           |          | generated always as (st_area(st_transform(the_geom, 3410))) stored | plain    |             |              |
 project_id    | uuid             |           |          |                                                                    | plain    |             |              |
 the_geom_hash | bytea            |           |          | generated always as (digest(the_geom::bytea, 'md5'::text)) stored  | extended |             |              |
Partition key: LIST (type)
Indexes:
    "planning_units_geom_pkey" PRIMARY KEY, btree (id, type)
    "planning_units_geom_geom_idx" gist (the_geom)
    "planning_units_geom_the_geom_type_project_id_check" UNIQUE, btree (the_geom_hash, type, COALESCE(project_id, '00000000-0000-0000-0000-000000000000'::uuid))
Check constraints:
    "features_geometry_valid_check" CHECK (st_isvalid(the_geom))
    "planning_units_geom_no_null_project_id" CHECK (project_id <> '00000000-0000-0000-0000-000000000000'::uuid)
Partitions: planning_units_geom_from_shapefile FOR VALUES IN ('from_shapefile'),
            planning_units_geom_hexagon FOR VALUES IN ('hexagon'),
            planning_units_geom_irregular FOR VALUES IN ('irregular'),
            planning_units_geom_square FOR VALUES IN ('square')

-- apidb
marxan-api=# \dT+ planning_unit_grid_shape
List of data types
-[ RECORD 1 ]-----+-------------------------
Schema            | public
Name              | planning_unit_grid_shape
Internal name     | planning_unit_grid_shape
Size              | 4
Elements          | square                  +
                  | hexagon                 +
                  | from_shapefile
Owner             | marxan-api
Access privileges |
Description       |
```

and finally, we have two Enums in the code:
```
export enum PlanningUnitGridShape {
  square = 'square',
  hexagon = 'hexagon',
  fromShapefile = 'from_shapefile',
}
```
and
```

export enum ShapeType {
  Square = 'square',
  Hexagon = 'hexagon',
  Irregular = 'irregular',
  FromShapefile = 'from_shapefile',
}
```
I'd leave things as they are for this PR, but to avoid confusion, I think this is a low-hanging fruit for some cleanup of accrued technical debt... I am just not sure how to treat the `irregular` shape - if we think we will need it in the future let's leave it, but I would at least converge to a single enum, and add a comment about the fact that the irregular shape is only expected to be used in the future when we add blah blah blah (I'm sure I should know, but I can't remember 😬)

aagm reviewed (commented):
So irregular exists because there is a future request from the client that will provide irregular grids like watersheds basins as pu

hotzevzl commented:
> So irregular exists because there is a future request from the client that will provide irregular grids like watersheds basins as pu

ah, yes - I keep forgetting 🤦🏼 So I have documented the enum we have in the source tree for this, marking `irregular` as `@todo` (I couldn't spot a better jsdoc tag for this) and adding the explanation you gave above... Hopefully that'll be enough for my future self 😬 

I have also taken this chance to converge the two enums we had in the code (`ShapeType`, which was added for the geoprocessing service, and `PlanningUnitGridShape`, which was the initial one) into a single one (`PlanningUnitGridShape`), and I have also synced the values and names of the underlying PostgreSQL types... and added a note in the enum comment that changes to the enum should be reflected in *both* PG types. We should be all set! I'll wait for CI to pass and merge.

I have also updated `mapshaper` to the latest version as we discussed. In practice, I am actually still seeing the issue that was puzzling me, though! Uppercase-extension `DBF` file (otherwise byte-identical to a lowercase-extension `dbf` one in a different test file) results in a different kind of GeoJSON once processed by mapshaper:

lowercase `.dbf` in zip file:

```
marxan-geoprocessing  | [Nest] 1891   - 02/21/2022, 3:51:24 PM   [ShapefileService] 698191c4-ea63-4f05-8b54-77763dc6c3ce_okavango_planning-grid.zip extracted successfully
marxan-geoprocessing  | geoJson.type: FeatureCollection
```

uppercase `.DBF` in zip file:

```
marxan-geoprocessing  | [Nest] 1891   - 02/21/2022, 3:51:57 PM   [ShapefileService] 21aa7d8b-0b6e-47ab-903b-56914f195902_okavango_planning-grid_UPPERCASE.zip extracted successfully
marxan-geoprocessing  | geoJson.type: GeometryCollection
marxan-geoprocessing  | [Nest] 1891   - 02/21/2022, 3:51:59 PM   [ShapefileService] Error: Types not supported +2583ms
marxan-geoprocessing  | [Nest] 1891   - 02/21/2022, 3:51:59 PM   [PlanningAreaController] failed to create a planning area from grid shapefile: Invalid Shapefile: Error: Types not supported +20ms
```

and for even more fun, a zip file that contains two identical dbf files, one with `.dbf` and one with `.DBF` extension:

```
marxan-geoprocessing  | [Nest] 1891   - 02/21/2022, 3:52:17 PM   [ShapefileService] fce38f31-5439-475c-9cb4-b846fa157036_okavango_planning-grid_AlLCaSe.zip extracted successfully
marxan-geoprocessing  | geoJson.type: FeatureCollection
```

I'm officially even more puzzled. But this will be a fight for another day 😢
